### PR TITLE
`Development`: Improve EnforceRoleInResourceAspects

### PR DESCRIPTION
**Pull request by MaximilianAnzinger**
#### General

- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.

### Description

Improves code design of enforce role in resource aspects by creating an abstract `EnforceRoleInResourceAspect` class.
Stacked on #8063

#### Performance Review
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [x] Code Review 2

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [AnnotationUtils.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6226/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.security.annotations/AnnotationUtils.html) | 85% | ✅ ❌ |
| [EnforceRoleInResourceAspect.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6226/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.security.annotations/EnforceRoleInResourceAspect.html) | 93% | ✅ ❌ |
| [EnforceRoleInCourseAspect.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6226/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.security.annotations.enforceRoleInCourse/EnforceRoleInCourseAspect.html) | 93% | ✅ ❌ |

## Summary by CodeRabbit

- **New Features**
	- Introduced an aspect for enforcing role-based authorization checks across the application, enhancing security by ensuring users have the required roles for accessing specific resources.

- **Refactor**
	- Updated annotation parameters for improved clarity and consistency, changing `courseIdFieldName` to `resourceIdFieldName` across several security-related annotations.
	- Enhanced type safety and error handling in utility methods through added parameter annotations.

- **Bug Fixes**
	- Adjusted test cases to reflect the updated annotation parameter naming, ensuring alignment with the new field naming convention and maintaining test accuracy.

### CODE REVIEW

**Review by julian-christl — APPROVED** [review]
Code looks good

File: src/main/java/de/tum/in/www1/artemis/security/annotations/EnforceRoleInResourceAspect.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+     * method if the user has the required role. Will otherwise return forbidden (as response entity)
+     *
+     * @param joinPoint Proceeding join point of the aspect
+     * @return The original return value of the called method, if the authorization checks pass, a forbidden response entity otherwise
+     * @throws Throwable If there was any error during method execution (both the aspect or the actual called method)
+     */
+    @Around(value = "callAt()", argNames = "joinPoint")
+    public Object around(ProceedingJoinPoint joinPoint) throws Throwable {
+        final var genericAnnotation = getAnnotation(genericAnnotationClass, joinPoint).orElseThrow();
+        final var role = getRole(genericAnnotation);
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/security/annotations/EnforceRoleInResourceAspect.java:85]: Should this also throw an exception if it does not exist?

File: src/main/java/de/tum/in/www1/artemis/security/annotations/enforceRoleInCourse/EnforceRoleInCourseAspect.java [hunk 2]
```diff
@@ -1,72 +1,36 @@
...
-    @Around(value = "callAt()", argNames = "joinPoint")
-    public Object around(ProceedingJoinPoint joinPoint) throws Throwable {
-        final var annotation = getAnnotation(EnforceRoleInCourse.class, joinPoint).orElseThrow();
-        final var courseIdFieldName = getCourseIdFieldName(annotation, joinPoint);
-        final var courseId = getIdFromSignature(joinPoint, courseIdFieldName).orElseThrow(() -> new IllegalArgumentException(
-                "Method annotated with @EnforceRoleInCourse must have a parameter named " + annotation.courseIdFieldName() + " of type long/Long."));
-        authorizationCheckService.checkIsAtLeastRoleInCourseElseThrow(annotation.value(), courseId);
-        return joinPoint.proceed();
+    @Override
+    protected void callAt() {
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/security/annotations/enforceRoleInCourse/EnforceRoleInCourseAspect.java:29]: Is this intended to do nothing?

**Review by rstief — APPROVED** [review]
Code LGTM, 2 small notes

File: src/main/java/de/tum/in/www1/artemis/security/annotations/enforceRoleInCourse/EnforceRoleInCourseAspect.java (hunk 2 above)
Comment (MaximilianAnzinger) [reply src/main/java/de/tum/in/www1/artemis/security/annotations/enforceRoleInCourse/EnforceRoleInCourseAspect.java:29]: Yes, this is just the point cut. For most purposes (including ours) this method should be empty.

**Pull request merged**